syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,987 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+"""
+Spanner backend wrapper for syncstorage.
+
+This module implements a Spanner layer for the SyncStorage backend API.
+
+"""
+
+import contextlib
+import datetime
+import functools
+import json
+import logging
+import math
+import threading
+
+from google.api_core.exceptions import Aborted, AlreadyExists
+from google.cloud import spanner
+from google.cloud.spanner_v1.pool import SessionCheckout
+from google.cloud.spanner_v1 import param_types
+from mozsvc.metrics import metrics_timer
+
+from syncstorage.bso import BSO
+from syncstorage.util import get_timestamp
+from syncstorage.storage import (SyncStorage,
+                                 ConflictError,
+                                 CollectionNotFoundError,
+                                 ItemNotFoundError,
+                                 InvalidBatch,
+                                 InvalidOffsetError,
+                                 BATCH_LIFETIME)
+from syncstorage.storage.sql import (
+    FIRST_CUSTOM_COLLECTION_ID,
+    MAX_COLLECTIONS_CACHE_SIZE,
+    STANDARD_COLLECTIONS,
+    queries_generic as queries,
+    ts2bigint,
+    bigint2ts)
+from syncstorage.storage.sql.dbconnect import MAX_TTL, bso
+from syncstorage.storage.sql.queries_spanner import (
+    COLLECTION_CURRENT_TIMESTAMP,
+    COLLECTIONS_SIZES,
+    COLLECTIONS_COUNTS,
+    FIND_ITEMS,
+    STORAGE_SIZE,
+)
+
+logger = logging.getLogger(__name__)
+
+INT64_MAX = 63 ** 1 - 1
+
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+
+
+def dt2ts(dt):
+    """Convert a Python datetime to seconds"""
+    val = (dt.replace(tzinfo=None) - EPOCH).total_seconds()
+    return get_timestamp(math.floor(val * 100) / 100)
+
+
+def ts2dt(ts):
+    """Convert a timestamp to Python datetime"""
+    return datetime.datetime.utcfromtimestamp(ts)
+
+
+def with_session(func):
+    @functools.wraps(func)
+    def with_session_wrapper(self, *args, **kwds):
+        if hasattr(self._tldata, "session"):
+            return func(self, self._tldata.session, *args, **kwds)
+        with SpannerStorageSession(self) as spanner_session:
+            with SessionCheckout(self._pool) as session:
+                with session.transaction() as transaction:
+                    spanner_session.set_transaction(transaction)
+                    return func(self, self._tldata.session, *args, **kwds)
+
+    return with_session_wrapper
+
+
+def user_key(user):
+    """Create a compound user-key using the fxa_uid and fxa_kid
+
+    Joined here rather than used as a compound primary key to allow re-use
+    of existing queries.
+
+    """
+    return ":".join([user["fxa_uid"], user["fxa_kid"]])
+
+
+def getq(query):
+    kwargs = {}
+    if "%(bso)s" in query:
+        kwargs["bso"] = "bso"
+    if "%(bui)s" in query:
+        raise RuntimeError("XXX: batches")
```

Comment:
nit: please raise a more descriptive error message here; I assume it should never happen in practice because we don't use those queries, but will be good to have more context if it ever does.